File: Dddml.Wms.Services.Tests/ShipmentTests.cs
using Dddml.Wms.Domain.AttributeSetInstance;
using Dddml.Wms.Domain.Product;
using Dddml.Wms.Domain.Shipment;
using Dddml.Wms.Domain.ShipmentType;
using Dddml.Wms.Domain.StatusItem;
using Dddml.Wms.Specialization;
using NUnit.Framework;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Dddml.Wms.Services.Tests
{
    public class ShipmentTests : TestsBase
    {
        IShipmentApplicationService shipmentApplicationService;

        IProductApplicationService productApplicationService;

        IAttributeSetInstanceApplicationService attributeSetInstanceApplicationService;

        [SetUp]
        public void SetUp()
        {
            base.SetUp();

            shipmentApplicationService = ApplicationContext.Current["shipmentApplicationService"] as IShipmentApplicationService;

            productApplicationService = ApplicationContext.Current["productApplicationService"] as IProductApplicationService;

            attributeSetInstanceApplicationService = ApplicationContext.Current["attributeSetInstanceApplicationService"] as IAttributeSetInstanceApplicationService;
        }

        [Test]
        public void TestImportAndReceive()
        {
            // ///////////////////////////////////////
            // Create a product for test.
            var prd_1 = CreateProduct_1();

            // ///////////////////////////////////////
            // Create a shipment.
            var shipmentId = CreateShipment_1(prd_1);

            UpdateShipmentToPurchShipShipped(shipmentId);

            ReceiveAllItems(shipmentId);

            UpdateShipmentToPurchShipReceived(shipmentId);
        }

        private void ReceiveAllItems(string shipmentId)
        {
            var shipmentState = shipmentApplicationService.Get(shipmentId);
            var version = shipmentState.Version;
            foreach (var item in shipmentState.ShipmentItems)
            {
                var receiveItem = new ShipmentCommands.ReceiveItem();
                receiveItem.ShipmentId = shipmentId;
                receiveItem.ShipmentItemSeqId = item.ShipmentItemSeqId;
                receiveItem.AcceptedQuantity = item.Quantity;
                receiveItem.RejectedQuantity = 0;
                receiveItem.DamagedQuantity = 0;

                if (!String.IsNullOrWhiteSpace(item.AttributeSetInstanceId))
                {
                    var srcAttrSetInst = attributeSetInstanceApplicationService.Get(item.AttributeSetInstanceId);
                    var rollId = srcAttrSetInst.SerialNumber;

                    var attrSetInst_1 = new Dictionary<string, object>();
                    // //////////////////////////////////
                    attrSetInst_1.Add("SerialNumber", rollId);
                    //attrSetInst_1.Add("WidthInch", (decimal)17.75);
                    //attrSetInst_1.Add("DiameterInch", 48.00);
                    //attrSetInst_1.Add("WeightLbs", (decimal)1678);
                    attrSetInst_1.Add("WeightKg", (TestWeightKg - (decimal)0.001));
                    //attrSetInst_1.Add("AirDryWeightLbs", (decimal)1705.682);
                    //attrSetInst_1.Add("AirDryWeightKg", (decimal)773.684);
                    //attrSetInst_1.Add("AirDryMetricTon", (decimal)0.774);
                    //attrSetInst_1.Add("PackageCount", 2);
                    //attrSetInst_1.Add("AirDryPct", (decimal)101.650);
                    // //////////////////////////////////
                    receiveItem.AttributeSetInstance = attrSetInst_1;
                }

                receiveItem.CommandId = Guid.NewGuid().ToString();
                receiveItem.Version = version;
                shipmentApplicationService.When(receiveItem);

                version++;
            }
        }

        static decimal TestWeightKg
        {
            get { return (decimal)761.125; }
        }

        private void UpdateShipmentToPurchShipShipped(string shipmentId)
        {
            //var updateShipment = new MergePatchShipment();
            var updateShipment = new ShipmentCommands.Ship();
            updateShipment.ShipmentId = shipmentId;
            updateShipment.CommandId = Guid.NewGuid().ToString();
            //updateShipment.StatusId = StatusItemIds.PurchShipShipped;
            updateShipment.Version = 1;
            shipmentApplicationService.When(updateShipment);
        }

        private void UpdateShipmentToPurchShipReceived(string shipmentId)
        {
            //var updateShipment = new MergePatchShipment();
            var updateShipment = new ShipmentCommands.ConfirmAllItemsReceived();
            updateShipment.ShipmentId = shipmentId;
            updateShipment.CommandId = Guid.NewGuid().ToString();
            //updateShipment.StatusId = StatusItemIds.PurchShipShipped;
            updateShipment.Version = 4; //todo???
            shipmentApplicationService.When(updateShipment);
        }

        private string CreateShipment_1(CreateProduct prd_1)
        {
            var shipmentId = DateTime.Now.Ticks.ToString();
            var shipImport = new ShipmentCommands.Import();
            shipImport.ShipmentId = shipmentId;

            var prdId = prd_1.ProductId;

            var rollId_1 = "H71051402A";
            var shipItem_1 = NewImportingShipmentItem(prdId, rollId_1);

            var rollId_2 = "H00000000A";
            var shipItem_2 = NewImportingShipmentItem(prdId, rollId_2);

            shipImport.ShipmentItems = new HashSet<ImportingShipmentItem>(new ImportingShipmentItem[] {
                    shipItem_1,
                    shipItem_2,
                });

            shipImport.ShipmentTypeId = ShipmentTypeIds.IncomingShipment;
            shipImport.DestinationFacilityId = "TEST_1";

            shipmentApplicationService.When(shipImport);
            return shipmentId;
        }

        private CreateProduct CreateProduct_1()
        {
            var prd_1 = new CreateProduct();
            prd_1.ProductId = DateTime.Now.Ticks.ToString();//Guid.NewGuid().ToString();
            prd_1.ProductName = "Test_" + prd_1.ProductId;
            prd_1.AttributeSetId = typeof(InitAttributeSets.FluffPulpAttrSetInstance).Name;
            prd_1.IsSerialNumbered = true;
            //prd_1.QuantityUomId;
            prd_1.CommandId = Guid.NewGuid().ToString();

            productApplicationService.When(prd_1);
            return prd_1;
        }

        private static ImportingShipmentItem NewImportingShipmentItem(string productId, string rollId)
        {
            var shipItem_1 = new ImportingShipmentItem();
            shipItem_1.ProductId = productId;
            shipItem_1.Quantity = 1;
            var attrSetInst_1 = new Dictionary<string, object>();

            // //////////////////////////////////
            attrSetInst_1.Add("SerialNumber", rollId);
            attrSetInst_1.Add("WidthInch", (decimal)17.75);
            attrSetInst_1.Add("DiameterInch", 48.00);
            attrSetInst_1.Add("WeightLbs", (decimal)1678);
            attrSetInst_1.Add("WeightKg", TestWeightKg);
            attrSetInst_1.Add("AirDryWeightLbs", (decimal)1705.682);
            attrSetInst_1.Add("AirDryWeightKg", (decimal)773.684);
            attrSetInst_1.Add("AirDryMetricTon", (decimal)0.774);
            attrSetInst_1.Add("PackageCount", 2);
            attrSetInst_1.Add("AirDryPct", (decimal)101.650);
            // //////////////////////////////////

            shipItem_1.AttributeSetInstance = attrSetInst_1;
            return shipItem_1;
        }

    }
}

File: java/Dddml.Wms.JavaServices/src/test/java/org/dddml/wms/ShipmentTests.java
package org.dddml.wms;

import org.dddml.wms.domain.attributesetinstance.AttributeSetInstanceApplicationService;
import org.dddml.wms.domain.attributesetinstance.AttributeSetInstanceState;
import org.dddml.wms.domain.product.AbstractProductCommand;
import org.dddml.wms.domain.product.ProductApplicationService;
import org.dddml.wms.domain.product.ProductCommand;
import org.dddml.wms.domain.service.OrderShipGroupApplicationService;
import org.dddml.wms.domain.service.OrderShipGroupServiceCommands;
import org.dddml.wms.domain.shipment.*;
import org.dddml.wms.domain.shipmenttype.ShipmentTypeIds;
import org.dddml.wms.specialization.ApplicationContext;

import java.math.BigDecimal;
import java.util.UUID;

/**
 * Created by yangjiefeng on 2018/2/4.
 */
public class ShipmentTests {
    private ShipmentApplicationService shipmentApplicationService;

    private ProductApplicationService productApplicationService;

    private AttributeSetInstanceApplicationService attributeSetInstanceApplicationService;

    private OrderShipGroupApplicationService orderShipGroupApplicationService;

    public void setUp() {
        //super.SetUp();
        shipmentApplicationService = (ShipmentApplicationService) ApplicationContext.current.get("shipmentApplicationService");
        productApplicationService = (ProductApplicationService) ApplicationContext.current.get("productApplicationService");
        attributeSetInstanceApplicationService = (AttributeSetInstanceApplicationService) ApplicationContext.current.get("attributeSetInstanceApplicationService");
        orderShipGroupApplicationService = (OrderShipGroupApplicationService) ApplicationContext.current.get("orderShipGroupApplicationService");
    }

    public void testImportAndReceive() {
        // ///////////////////////////////////////
        // Create a product for test.
        ProductCommand.CreateProduct prd_1 = createProduct_1();

        // ///////////////////////////////////////
        // Create a shipment.
        String shipmentId = createShipment_1(prd_1);

        updatePurchaseShipmentToShipped(shipmentId);

        receiveAllItems(shipmentId);

        updateShipmentToPurchShipReceived(shipmentId);
    }

    private static java.math.BigDecimal getTestWeightKg() {
        return java.math.BigDecimal.valueOf(761.125);
    }

    private void updateShipmentToPurchShipReceived(String shipmentId) {
        ShipmentState shipmentState = shipmentApplicationService.get(shipmentId);
        //var updateShipment = new MergePatchShipment();
        ShipmentCommands.ConfirmAllItemsReceived updateShipment = new ShipmentCommands.ConfirmAllItemsReceived();
        updateShipment.setShipmentId(shipmentId);
        updateShipment.setVersion(shipmentState.getVersion());
        updateShipment.setDestinationLocatorId(InOutTests.TEST_LOCATOR_ID_1_RECEIVING_AREA);
        updateShipment.setCommandId(UUID.randomUUID().toString());
        //updateShipment.StatusId = StatusItemIds.PurchShipShipped;
        shipmentApplicationService.when(updateShipment);
    }

    private void receiveAllItems(String shipmentId) {
        ShipmentState shipmentState = shipmentApplicationService.get(shipmentId);
        Long version = shipmentState.getVersion();
        int i = 0;
        for (ShipmentItemState item : shipmentState.getShipmentItems()) {
            ShipmentCommands.ReceiveItem receiveItem = new ShipmentCommands.ReceiveItem();
            receiveItem.setShipmentId(shipmentId);
            receiveItem.setShipmentItemSeqId(item.getShipmentItemSeqId());
            receiveItem.setAcceptedQuantity(item.getQuantity());
            receiveItem.setRejectedQuantity(BigDecimal.ZERO);
            receiveItem.setDamagedQuantity(BigDecimal.ZERO);

            if (!(null == item.getAttributeSetInstanceId())) {
                if (i == 0) {

                    AttributeSetInstanceState srcAttrSetInst = attributeSetInstanceApplicationService.get(item.getAttributeSetInstanceId());
                    String rollId = srcAttrSetInst.getSerialNumber();

                    java.util.HashMap<String, Object> attrSetInst_1 = new java.util.HashMap<String, Object>();
                    // //////////////////////////////////
                    attrSetInst_1.put("SerialNumber", rollId);
                    //attrSetInst_1.Add("WidthInch", (decimal)17.75);
                    //attrSetInst_1.Add("DiameterInch", 48.00);
                    //attrSetInst_1.Add("WeightLbs", (decimal)1678);
                    attrSetInst_1.put("WeightKg",
                            getTestWeightKg().subtract(java.math.BigDecimal.valueOf(0.001)));
                    //attrSetInst_1.Add("AirDryWeightLbs", (decimal)1705.682);
                    //attrSetInst_1.Add("AirDryWeightKg", (decimal)773.684);
                    //attrSetInst_1.Add("AirDryMetricTon", (decimal)0.774);
                    //attrSetInst_1.Add("RollCnt", 2);
                    //attrSetInst_1.Add("AirDryPct", (decimal)101.650);
                    // //////////////////////////////////
                    receiveItem.setAttributeSetInstance(attrSetInst_1);

                } else {

                    AttributeSetInstanceState srcAttrSetInst = attributeSetInstanceApplicationService.get(item.getAttributeSetInstanceId());
                    String attrSetInstId = srcAttrSetInst.getAttributeSetInstanceId();
                    java.util.HashMap<String, Object> attrSetInst_2 = new java.util.HashMap<String, Object>();
                    attrSetInst_2.put("attributeSetInstanceId", attrSetInstId);
                    receiveItem.setAttributeSetInstance(attrSetInst_2);

                }
                i++;
            }
            // else {
            //    throw new RuntimeException("null == item.getAttributeSetInstanceId()");
            //}

            receiveItem.setCommandId(UUID.randomUUID().toString());
            receiveItem.setVersion(version);
            receiveItem.setReceiptSeqId(receiveItem.getShipmentItemSeqId());
            shipmentApplicationService.when(receiveItem);

            version++;
        }
    }

    private void updatePurchaseShipmentToShipped(String shipmentId) {
        //long firstVersion = 0L;
        //var updateShipment = new MergePatchShipment();
        OrderShipGroupServiceCommands.ShipPOShipment updateShipment = new OrderShipGroupServiceCommands.ShipPOShipment();
        updateShipment.setShipmentId(shipmentId);
        updateShipment.setCommandId(UUID.randomUUID().toString());
        updateShipment.setHintShipmentItemsEnabled(true);
        //updateShipment.StatusId = StatusItemIds.PurchShipShipped;
        //updateShipment.setVersion(firstVersion);
        orderShipGroupApplicationService.when(updateShipment);
    }

    private String createShipment_1(ProductCommand.CreateProduct prd_1) {
        String shipmentId = new java.util.Date().getTime() + "";
        ShipmentCommands.Import shipImport = new ShipmentCommands.Import();
        shipImport.setShipmentId(shipmentId);

        String prdId = prd_1.getProductId();

        String rollId_1 = "" + UUID.randomUUID().hashCode();//"H71051402A";
        ImportingShipmentItem shipItem_1 = newImportingShipmentItem(prdId, rollId_1);

        String rollId_2 = "" + UUID.randomUUID().hashCode();//"H00000000A";
        ImportingShipmentItem shipItem_2 = newImportingShipmentItem(prdId, rollId_2);

        shipImport.setShipmentItems(
                        java.util.Arrays.asList(new ImportingShipmentItem[]{shipItem_1, shipItem_2})
        );

        //shipImport.setShipmentTypeId(ShipmentTypeIds.INCOMING_SHIPMENT);
        shipImport.setShipmentTypeId(ShipmentTypeIds.PURCHASE_SHIPMENT);//先保证采购装运是正确的。
        shipImport.setDestinationFacilityId("TEST_1");
        shipImport.setPrimaryOrderId("" + UUID.randomUUID().hashCode());

        shipmentApplicationService.when(shipImport);
        return shipmentId;
    }

    private static ImportingShipmentItem newImportingShipmentItem(String productId, String rollId) {
        ImportingShipmentItem shipItem_1 = new ImportingShipmentItem();
        shipItem_1.setProductId(productId);
        shipItem_1.setQuantity(BigDecimal.valueOf(1));
        java.util.HashMap<String, Object> attrSetInst_1 = new java.util.HashMap<String, Object>();

        // //////////////////////////////////
        attrSetInst_1.put("SerialNumber", rollId);
        attrSetInst_1.put("WidthInch", java.math.BigDecimal.valueOf(17.75));
        attrSetInst_1.put("DiameterInch", 48.00);
        attrSetInst_1.put("WeightLbs", 1678);
        attrSetInst_1.put("WeightKg", getTestWeightKg());
        attrSetInst_1.put("AirDryWeightLbs", java.math.BigDecimal.valueOf(1705.682));
        attrSetInst_1.put("AirDryWeightKg", java.math.BigDecimal.valueOf(773.684));
        attrSetInst_1.put("AirDryMetricTon", java.math.BigDecimal.valueOf(0.774));
        attrSetInst_1.put("RollCnt", 2);
        attrSetInst_1.put("AirDryPct", java.math.BigDecimal.valueOf(101.650));
        // //////////////////////////////////

        shipItem_1.setAttributeSetInstance(attrSetInst_1);
        return shipItem_1;
    }


    private ProductCommand.CreateProduct createProduct_1() {
        ProductCommand.CreateProduct prd_1 = new AbstractProductCommand.SimpleCreateProduct();
        prd_1.setProductId(new java.util.Date().getTime() + ""); //Guid.NewGuid().ToString();
        prd_1.setProductName("Test_" + prd_1.getProductId());
        prd_1.setAttributeSetId(InitAttributeSets.FLUFF_PULP_ATTR_SET_ID);//"FluffPulpAttrSet");
        prd_1.setIsSerialNumbered(true);
        //prd_1.QuantityUomId;
        prd_1.setCommandId(UUID.randomUUID().toString());

        productApplicationService.when(prd_1);
        return prd_1;
    }

}
